prereise/gather/demanddata/transportation_electrification/smart_charging.py

```diff
@@ -0,0 +1,365 @@
+import dwelling
+import numpy as np
+import scipy
+
+
+def smart_charging(
+    census_region,
+    model_year,
+    veh_range,
+    kwhmi,
+    power,
+    location_strategy,
+    veh_type,
+    filepath,
+    comm_type,
+    trip_strategy=1,
+):
+    """Smart charging function
+
+    :param int census_region: any of the 9 census regions defined by US census bureau.
+    :param int model_year: year that is being modelled/projected to, 2017, 2030, 2040, 2050.
+    :param int veh_range: 100, 200, or 300, represents how far vehicle can travel on single charge.
+    :param int kwhmi: fuel efficiency, should vary based on vehicle type and model_year.
+    :param int power: charger power, EVSE kW.
+    :param int location_strategy: where the vehicle can charge-1, 2, 3, 4, or 5; 1-home only, 2-home and work related, 3-anywhere if possibile, 4-home and school only, 5-home and work and school.
+    :param str veh_type: determine which category (LDV or LDT) to produce charging profiles for
+    :param str filepath: the path to the nhts mat file.
+    :param int comm_type: the type of Commute
+    :param int trip_strategy: determine to charge after any trip (1) or only after the last trip (2)
+    :return: (*numpy.ndarray*) -- charging profiles.
+    """
+
+    # load NHTS data from function
+    if veh_type.lower() == "ldv":
+        newdata = data_helper.remove_ldt(data_helper.load_data(census_region, filepath))
+
+    elif veh_type.lower() == "ldt":
+        newdata = data_helper.remove_ldv(data_helper.load_data(census_region, filepath))
+
+    # updates the weekend and weekday values in the nhts data
+    newdata = data_helper.update_if_weekend(newdata)
+
+    new_columns = [
+        "Electricity cost",
+        "Battery discharge",
+        "Battery charge",
+        "Battery size",
+    ]
+    newdata = newdata.reindex(list(newdata.columns) + new_columns, axis=1, fill_value=0)
+
+    input_day = data_helper.get_input_day(data_helper.get_model_year_dti(model_year))
+
+    TRANS_charge = np.zeros(24 * len(input_day))
+    data_day = data_helper.get_data_day(newdata)
+
+    daily_vmt_total = data_helper.total_daily_vmt(
+        census_region, comm_type, location_strategy, input_day
+    )
+
+    kwh = kwhmi * veh_range
+    emfacvmt = 758118400
+
+    LOAD = np.ones(24 * len(input_day))
+
+    nd_len = len(newdata)
+
+    for day_iter in range(len(input_day)):
+
+        if day_iter == len(input_day) - 1:
+            LOAD_adj = [
+                LOAD[i] + TRANS_charge[i]
+                for i in range(day_iter * 24, (day_iter + 1) * 24)
+            ]
+            LOAD_adj += [LOAD[i] + TRANS_charge[i] for i in range(24)]
+        else:
+            LOAD_adj = [
+                LOAD[i] + TRANS_charge[i]
+                for i in range(day_iter * 24, day_iter * 24 + 48)
+            ]
+
+        cost = LOAD_adj
+
+        G2Vload = np.zeros((100, 48))
+        individualG2Vload = np.zeros(nd_len, 48)
+
+        i = 0
+
+        while i < nd_len:
```

Comment:
This nested loop is probably the source of the slow performance, since we end up with (number_of_trips * number_of_days_in_year) iterations, and we perform an optimization solve within each one. Are we performing any repeated calculations within here? Can any of the calculations be batched to work on a group of trips, rather than individual ones?